src/main/java/org/broadinstitute/hellbender/engine/LocusWalkerByInterval.java

```diff
@@ -0,0 +1,145 @@
+package org.broadinstitute.hellbender.engine;
+
+import htsjdk.samtools.SAMFileHeader;
+import htsjdk.samtools.SAMReadGroupRecord;
+import htsjdk.samtools.util.Locatable;
+import htsjdk.samtools.util.OverlapDetector;
+import org.apache.commons.collections4.SetUtils;
+import org.broadinstitute.hellbender.engine.filters.CountingReadFilter;
+import org.broadinstitute.hellbender.utils.SimpleInterval;
+import org.broadinstitute.hellbender.utils.locusiterator.AlignmentContextIteratorBuilder;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * An implementation of {@link LocusWalker} that supports arbitrary side locus side inputs as a track.
+ *
+ * The class works as follows: Before traversal {@link #getIntervalObjectsToQueryOver()} will be called to generate a global
+ * list of Locatable that will be stored in memory for the duration of the traversal. For each site {@link #apply(AlignmentContext, ReferenceContext, FeatureContext, Set)}} will be called
+ * with a set consisting of all the provided overlapping intervals. The first time any Locus overlaps with the traversal
+ * {@link #onIntervalStart(Locatable)} will be called once. Once an Locatable has been passed by the traversal {@link #onIntervalEnd(Locatable)}
+ * will be called once. Otherwise See {@link LocusWalker} for general implementation details.
+ *
+ * NOTE: If there are Loctables provided by {@link #getIntervalObjectsToQueryOver()} that are never covered by the traversal of
+ * the tool, {@link #onIntervalStart(Locatable)} and {@link #onIntervalEnd(Locatable)} will not be called.
+ */
+public abstract class LocusWalkerByInterval extends LocusWalker {
+
+    private OverlapDetector<Locatable> intervalsToTrack = null;
+    private Set<Locatable> previousIntervals = new HashSet<>();
```

Comment:
So as it is currently written the consequence here is very low. Auditing this one the worst case here is that the order in which one writes out genes could be shuffled somewhat (the other cases were order independent). I'm going to switch this to a linked hash set to avoid the trouble though.